src/Compilers/CSharp/Portable/Symbols/Metadata/PE/NativeIntegerTypeDecoder.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Reflection.Metadata;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
+{
+    internal struct NativeIntegerTypeDecoder
+    {
+        internal static TypeSymbol TransformType(TypeSymbol type, EntityHandle handle, PEModuleSymbol containingModule)
+        {
+            return containingModule.Module.HasNativeIntegerAttribute(handle, out var transformFlags) ?
+                TransformType(type, transformFlags) :
+                type;
+        }
+
+        private static TypeSymbol TransformType(TypeSymbol type, ImmutableArray<bool> transformFlags)
+        {
+            var decoder = new NativeIntegerTypeDecoder(transformFlags);
+            try
+            {
+                var result = decoder.TransformType(type);
+                if (decoder._index == transformFlags.Length)
+                {
+                    return result;
+                }
+            }
+            catch (ArgumentException)
+            {
+            }
+            return type;
+        }
+
+        private readonly ImmutableArray<bool> _transformFlags;
+        private int _index;
+
+        private NativeIntegerTypeDecoder(ImmutableArray<bool> transformFlags)
+        {
+            _transformFlags = transformFlags;
+            _index = 0;
+        }
+
+        private TypeWithAnnotations TransformTypeWithAnnotations(TypeWithAnnotations type)
+        {
+            return type.WithTypeAndModifiers(TransformType(type.Type), type.CustomModifiers);
+        }
+
+        private TypeSymbol TransformType(TypeSymbol type)
+        {
+            return type switch
+            {
+                NamedTypeSymbol namedType => TransformNamedType(namedType),
+                ArrayTypeSymbol arrayType => TransformArrayType(arrayType),
+                PointerTypeSymbol pointerType => TransformPointerType(pointerType),
+                _ => type,
```

Comment:
>type [](start = 21, length = 4)

What is getting here? I think it would be better to explicitly handle what we know today, and at least assert for other types.